### PR TITLE
Added the Combat Bakery Kit

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Adds the Combat Bakery Kit. Contains an edible weapon disguised as a baguette, which hits for 20 sharp damage and can block, a pair of throwing croissants, which can be ate to conceal evidence, as well as boomeranging back to the owner on throw.
Finally, a book that teaches them how to make these deadly throwing croissants. Anyone can use the Croissants, but only people that read this book can craft them, and it has one use, unless one can find a way to recharge it. 25 tc.

Ports the Boomerang component from TG. Does what it says on the tin.

Ports the granter item, and recipe granter items from TG. TG uses this system to replace book of babel / spellbooks / cqc spawners and all that, would be good to convert to this system in the future.

## Why It's Good For The Game

A fun kit to kill people in a delicious way is good, as well as the fact you can literally eat the evidence of your actions! Reasonably priced at 5 TC, one can get a sharpened knife as chef very easily, and while throwing stars are nice, unlike the throwing kit they do not embed, and a stech will probably serve you better, but a stech may not be as fun!

## Testing

I went to the kitchen and I baked a croissant, and used the book from uplink to be able to craft it, then threw it at people and things.

## Changelog
:cl:
add: Added the Combat Bakery Kit. A kit of clandestine baked weapons. Contains a baguette which a skilled mime could use as a sword, a pair of throwing croissants, and the recipe to make more on demand. Once the job is done, eat the evidence.
/:cl:

### CODE REVIEW

File: code/datums/components/boomerang.dm [hunk 1]
```diff
@@ -0,0 +1,89 @@
+///The cooldown period between last_boomerang_throw and it's methods of implementing a rebound proc.
+#define BOOMERANG_REBOUND_INTERVAL (1 SECONDS)
+/**
```
Comment (lewcc) [inline code/datums/components/boomerang.dm:3]:
```suggestion

/**
```

File: code/datums/components/boomerang.dm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+/datum/component/boomerang/Initialize(boomerang_throw_range, thrower_easy_catch_enabled)
+	. = ..()
+	if(!isitem(parent)) //Only items support being thrown around like a boomerang, feel free to make this apply to humans later on.
+		return COMPONENT_INCOMPATIBLE
+
+	//Assignments
+	if(boomerang_throw_range)
+		src.boomerang_throw_range = boomerang_throw_range
+	if(thrower_easy_catch_enabled)
+		src.thrower_easy_catch_enabled = thrower_easy_catch_enabled
```
Comment (lewcc) [inline code/datums/components/boomerang.dm:24]: Should be no need to check the values like this unless you're transferring the component
```suggestion
	src.boomerang_throw_range = boomerang_throw_range
	src.thrower_easy_catch_enabled = thrower_easy_catch_enabled
```

File: code/datums/components/boomerang.dm [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ * * thrown_thing: The thrownthing datum from the parent object's latest throw. Updates thrown_boomerang.
+ * * spin: Carry over from POST_THROW, the speed of rotation on the boomerang when thrown.
+ */
+/datum/component/boomerang/proc/prepare_throw(datum/source, datum/thrownthing/thrown_thing, spin)
+	SIGNAL_HANDLER
+	if(thrower_easy_catch_enabled && thrown_thing?.thrower)
+		if(iscarbon(thrown_thing.thrower))
+			var/mob/living/carbon/Carbon = thrown_thing.thrower
+			Carbon.throw_mode_on()
+	return
```
Comment (lewcc) [inline code/datums/components/boomerang.dm:46]:
```suggestion
	if(thrower_easy_catch_enabled && iscarbon(thrown_thing?.thrower))
		var/mob/living/carbon/C = thrown_thing.thrower
		C.throw_mode_on()
```

File: code/game/objects/items/granters/_granters.dm [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+	if(!do_after(user, reading_time, src))
+		return FALSE
+
+	to_chat(user, "<span class='notice'>[length(remarks) ? pick(remarks) : "You keep reading..."]</span>")
+	return TRUE
+
+/// Effects that occur whenever the book is read when it has no uses left.
+/obj/item/book/granter/proc/recoil(mob/living/user)
+
```
Comment (lewcc) [inline code/game/objects/items/granters/_granters.dm:78]:
```suggestion
/obj/item/book/granter/proc/recoil(mob/living/user)
    return

```

File: code/game/objects/items/granters/crafting_granters/_crafting_granter.dm [hunk 5]
```diff
@@ -0,0 +1,18 @@
+/obj/item/book/granter/crafting_recipe
+	/// A list of all recipe types we grant on learn
+	var/list/crafting_recipe_types = list()
+
+/obj/item/book/granter/crafting_recipe/on_reading_finished(mob/user)
+	. = ..()
```
Comment (lewcc) [inline code/game/objects/items/granters/crafting_granters/_crafting_granter.dm:6]:
```suggestion
	..()
```

File: code/game/objects/items/granters/crafting_granters/combat_baking.dm [hunk 6]
```diff
@@ -0,0 +1,19 @@
+/obj/item/book/granter/crafting_recipe/combat_baking
+	name = "the anarchist's cookbook"
+	desc = "A widely illegal recipe book, which will teach you how to bake croissants to die for."
+	crafting_recipe_types = list(
+		/datum/crafting_recipe/throwing_croissant,
+	)
+	icon_state = "cooking_learing_illegal"
```
Comment (lewcc) [inline code/game/objects/items/granters/crafting_granters/combat_baking.dm:7]: doesn't matter too much, but `s/learing/learning`

File: code/game/objects/items/granters/crafting_granters/combat_baking.dm [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/obj/item/book/granter/crafting_recipe/combat_baking
+	name = "the anarchist's cookbook"
+	desc = "A widely illegal recipe book, which will teach you how to bake croissants to die for."
```
Comment (lewcc) [inline code/game/objects/items/granters/crafting_granters/combat_baking.dm:3]:
```suggestion
	desc = "A widely illegal recipe book which will teach you how to bake croissants to die for."
```

File: code/game/objects/items/granters/crafting_granters/combat_baking.dm [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+	remarks = list(
+		"\"Austrian? Not French?\"",
+		"\"Got to get the butter ratio right...\"",
+		"\"This is the greatest thing since sliced bread!\"",
+		"\"I'll leave no trace except crumbs!\"",
+		"\"Who knew that bread could hurt a man so badly?\"",
+	)
+
+/obj/item/book/granter/crafting_recipe/combat_baking/recoil(mob/living/user)
+	to_chat(user, "<span class='warning'>The book dissolves into burnt flour!")
```
Comment (lewcc) [inline code/game/objects/items/granters/crafting_granters/combat_baking.dm:17]:
```suggestion
	to_chat(user, "<span class='warning'>The book dissolves into burnt flour!</span>")
```

File: code/datums/uplink_items/uplink_traitor.dm [hunk 9]
```diff
@@ -58,6 +58,15 @@
 	cost = 10
 	job = list("Mime")
 
+/datum/uplink_item/jobspecific/combat_baking
+	name = "Combat Bakery Kit"
+	desc = "A kit of clandestine baked weapons. Contains a baguette which a skilled mime could use as a sword, \
+		a pair of throwing croissants, and the recipe to make more on demand. Once the job is done, eat the evidence."
+	reference = "CBK"
+	item = /obj/item/storage/box/syndie_kit/combat_baking
+	cost = 5 //A chef can get a knife that sharp easily, though it won't block. While you can get endless boomerang, they are less deadly than a stech, and slower / more predictable.
```
Comment (Burzah) [inline code/datums/uplink_items/uplink_traitor.dm:67]:
```suggestion
	cost = 25 //A chef can get a knife that sharp easily, though it won't block. While you can get endless boomerang, they are less deadly than a stech, and slower / more predictable.
```
To account for the new TC inflation.

**Review by Burzah — CHANGES_REQUESTED** [review]

File: code/datums/components/boomerang.dm [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * Proc'd before the first thrown is performed in order to gather information regarding each throw as well as handle throw_mode as necessary.
+ * * source: Datum src from original signal call.
+ * * thrown_thing: The thrownthing datum from the parent object's latest throw. Updates thrown_boomerang.
+ * * spin: Carry over from POST_THROW, the speed of rotation on the boomerang when thrown.
+ */
+/datum/component/boomerang/proc/prepare_throw(datum/source, datum/thrownthing/thrown_thing, spin)
+	SIGNAL_HANDLER
+	if(thrower_easy_catch_enabled && iscarbon(thrown_thing?.thrower))
```
Comment (lewcc) [inline code/datums/components/boomerang.dm:41]: Is it possible that thrown thing could even be null here?

File: code/datums/components/boomerang.dm [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	aerodynamic_swing(throwing_datum, true_parent)
+
+/**
+ * Proc that triggers when the thrown boomerang has been fully thrown, rethrowing the boomerang back to the thrower, and producing visible feedback.
+ * * throwing_datum: The thrownthing datum that originally impacted the object, that we use to build the new throwing datum for the rebound.
+ * * hit_atom: The atom that has been hit by the boomerang'd object.
+ */
+/datum/component/boomerang/proc/aerodynamic_swing(datum/thrownthing/throwing_datum, obj/item/true_parent)
+	var/mob/thrown_by = locateUID(true_parent.thrownby)
+	if(thrown_by)
```
Comment (lewcc) [inline code/datums/components/boomerang.dm:77]:
```suggestion
	if(istype(thrown_by))
```

File: code/datums/components/boomerang.dm (hunk 10 above)
Comment (Qwertytoforty) [reply code/datums/components/boomerang.dm:41]: Not sure, I presume TG had it for a reason?
Or maybe turrets throwing croissonts?

File: code/modules/crafting/recipes.dm [hunk 12]
```diff
@@ -67,6 +67,17 @@
 	category = CAT_WEAPONRY
 	subcategory = CAT_WEAPON
 
+/datum/crafting_recipe/throwing_croissant
+	name = "Throwing croissant"
+	reqs = list(
+		/obj/item/reagent_containers/food/snacks/croissant = 1,
+		/obj/item/stack/rods = 1,
+	)
```
Comment (Henri215) [inline code/modules/crafting/recipes.dm:75]:
```suggestion
		/obj/item/stack/rods = 1
	)
```

File: code/game/objects/items/granters/crafting_granters/combat_baking.dm [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+		/datum/crafting_recipe/throwing_croissant,
+	)
+	icon_state = "cooking_learing_illegal"
+	remarks = list(
+		"\"Austrian? Not French?\"",
+		"\"Got to get the butter ratio right...\"",
+		"\"This is the greatest thing since sliced bread!\"",
+		"\"I'll leave no trace except crumbs!\"",
+		"\"Who knew that bread could hurt a man so badly?\"",
+	)
```
Comment (Henri215) [inline code/game/objects/items/granters/crafting_granters/combat_baking.dm:14]:
```suggestion
		"\"Who knew that bread could hurt a man so badly?\""
	)
```

File: code/game/objects/items/granters/crafting_granters/combat_baking.dm [hunk 14]
```diff
@@ -0,0 +1,19 @@
+/obj/item/book/granter/crafting_recipe/combat_baking
+	name = "the anarchist's cookbook"
+	desc = "A widely illegal recipe book which will teach you how to bake croissants to die for."
+	crafting_recipe_types = list(
+		/datum/crafting_recipe/throwing_croissant,
+	)
```
Comment (Henri215) [inline code/game/objects/items/granters/crafting_granters/combat_baking.dm:6]:
```suggestion
		/datum/crafting_recipe/throwing_croissant
	)
```

File: code/game/objects/items/granters/_granters.dm [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+	/// The amount of uses on the granter.
+	var/uses = 1
+	/// The time it takes to read the book
+	var/reading_time = 5 SECONDS
+	/// The sounds played as the user's reading the book.
+	var/list/book_sounds = list(
+		'sound/effects/pageturn1.ogg',
+		'sound/effects/pageturn2.ogg',
+		'sound/effects/pageturn3.ogg',
+	)
```
Comment (Henri215) [inline code/game/objects/items/granters/_granters.dm:22]:
```suggestion
		'sound/effects/pageturn3.ogg'
	)
```

File: code/game/objects/items/granters/_granters.dm [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+	if(locate(granted_action) in user.actions)
+		to_chat(user, "<span class='warning'>You already know all about [action_name]!</span>")
+		return FALSE
+	return TRUE
+
+/obj/item/book/granter/action/on_reading_start(mob/living/user)
+	to_chat(user, "<span class='notice'>You start reading about [action_name]...</span>")
+
+/obj/item/book/granter/action/on_reading_finished(mob/living/user)
+	to_chat(user, "<span class='notice'>You feel like you've got a good handle on [action_name]!</span>")
```
Comment (lewcc) [inline code/game/objects/items/granters/_granters.dm:103]:
```suggestion
/obj/item/book/granter/action/on_reading_finished(mob/living/user)
	SHOULD_CALL_PARENT(TRUE)
	to_chat(user, "<span class='notice'>You feel like you've got a good handle on [action_name]!</span>")
```

**Review by lewcc — CHANGES_REQUESTED** [review]
Looks pretty good, just one more thing.

File: code/game/objects/items/granters/_granters.dm (hunk 16 above)
Comment (S34NW) [reply code/game/objects/items/granters/_granters.dm:103]: @Qwertytoforty please dont mark this as resolved without actually resolving it or saying anything

**Review by S34NW — APPROVED** [review]

**Pull request merged**